### PR TITLE
pkg: support generic for btree

### PR DESCRIPTION
**Pull request by rleungx**
Signed-off-by: Ryan Leung <rleungx@gmail.com>

### What problem does this PR solve?

Issue Number: Ref #5648 #5606.

### What is changed and how does it work?
```
cd pkg/btree
go test -bench . -benchmem -count 5 > master/generic.txt
benchstat master.txt generic.txt
```
```
name                                 old time/op    new time/op    delta
Insert-8                                138ns ± 1%     106ns ± 0%   -23.06%  (p=0.008 n=5+5)
Seek-8                                  113ns ± 0%      74ns ± 1%   -34.79%  (p=0.008 n=5+5)
DeleteInsert-8                          294ns ± 1%     227ns ± 1%   -23.05%  (p=0.008 n=5+5)
DeleteInsertCloneOnce-8                 295ns ± 2%     228ns ± 1%   -22.85%  (p=0.008 n=5+5)
DeleteInsertCloneEachTime-8            1.18µs ± 1%    0.72µs ± 2%   -39.07%  (p=0.008 n=5+5)
Delete-8                                159ns ± 1%     120ns ± 1%   -24.72%  (p=0.008 n=5+5)
Get-8                                   104ns ± 1%      68ns ± 1%   -34.42%  (p=0.008 n=5+5)
GetCloneEachTime-8                      152ns ± 1%     114ns ± 1%   -25.00%  (p=0.008 n=5+5)
Ascend-8                               39.9µs ± 0%    31.1µs ± 0%   -21.98%  (p=0.008 n=5+5)
Descend-8                              38.9µs ± 1%    30.1µs ± 0%   -22.65%  (p=0.008 n=5+5)
AscendRange-8                          71.7µs ± 0%    49.8µs ± 0%   -30.57%  (p=0.008 n=5+5)
DescendRange-8                         93.0µs ± 0%    62.0µs ± 1%   -33.27%  (p=0.008 n=5+5)
AscendGreaterOrEqual-8                 52.3µs ± 1%    39.3µs ± 1%   -24.91%  (p=0.008 n=5+5)
DescendLessOrEqual-8                   74.5µs ± 0%    51.4µs ± 1%   -30.99%  (p=0.008 n=5+5)
DeleteAndRestore/CopyBigFreeList-8     4.87ms ± 0%    3.70ms ± 0%   -24.11%  (p=0.008 n=5+5)
DeleteAndRestore/Copy-8                4.99ms ± 0%    3.72ms ± 0%   -25.36%  (p=0.016 n=4+5)
DeleteAndRestore/ClearBigFreelist-8    2.45ms ± 0%    1.85ms ± 0%   -24.78%  (p=0.016 n=5+4)
DeleteAndRestore/Clear-8               2.58ms ± 2%    1.87ms ± 0%   -27.24%  (p=0.008 n=5+5)

name                                 old alloc/op   new alloc/op   delta
Insert-8                                37.0B ± 0%     19.0B ± 0%      ~     (p=0.079 n=4+5)
Seek-8                                  7.00B ± 0%     0.00B       -100.00%  (p=0.008 n=5+5)
DeleteInsert-8                          0.00B          0.00B           ~     (all equal)
DeleteInsertCloneOnce-8                 0.00B          0.00B           ~     (all equal)
DeleteInsertCloneEachTime-8            3.61kB ± 1%    2.53kB ± 1%   -29.94%  (p=0.008 n=5+5)
Delete-8                                0.00B          0.00B           ~     (all equal)
Get-8                                   0.00B          0.00B           ~     (all equal)
GetCloneEachTime-8                      48.0B ± 0%     48.0B ± 0%      ~     (all equal)
Ascend-8                                0.00B          0.00B           ~     (all equal)
Descend-8                               0.00B          0.00B           ~     (all equal)
AscendRange-8                           0.00B          0.00B           ~     (all equal)
DescendRange-8                          0.00B          0.00B           ~     (all equal)
AscendGreaterOrEqual-8                  0.00B          0.00B           ~     (all equal)
DescendLessOrEqual-8                    0.00B          0.00B           ~     (all equal)
DeleteAndRestore/CopyBigFreeList-8      275kB ± 0%     143kB ± 0%   -47.94%  (p=0.008 n=5+5)
DeleteAndRestore/Copy-8                 884kB ± 0%     466kB ± 0%   -47.30%  (p=0.008 n=5+5)
DeleteAndRestore/ClearBigFreelist-8    1.30kB ± 0%    0.86kB ± 0%   -33.38%  (p=0.008 n=5+5)
DeleteAndRestore/Clear-8                562kB ± 0%     299kB ± 0%   -46.82%  (p=0.008 n=5+5)

name                                 old allocs/op  new allocs/op  delta
Insert-8                                 0.00           0.00           ~     (all equal)
Seek-8                                   0.00           0.00           ~     (all equal)
DeleteInsert-8                           0.00           0.00           ~     (all equal)
DeleteInsertCloneOnce-8                  0.00           0.00           ~     (all equal)
DeleteInsertCloneEachTime-8              13.0 ± 0%      13.0 ± 0%      ~     (all equal)
Delete-8                                 0.00           0.00           ~     (all equal)
Get-8                                    0.00           0.00           ~     (all equal)
GetCloneEachTime-8                       3.00 ± 0%      3.00 ± 0%      ~     (all equal)
Ascend-8                                 0.00           0.00           ~     (all equal)
Descend-8                                0.00           0.00           ~     (all equal)
AscendRange-8                            0.00           0.00           ~     (all equal)
DescendRange-8                           0.00           0.00           ~     (all equal)
AscendGreaterOrEqual-8                   0.00           0.00           ~     (all equal)
DescendLessOrEqual-8                     0.00           0.00           ~     (all equal)
DeleteAndRestore/CopyBigFreeList-8       15.0 ± 0%      14.0 ± 0%    -6.67%  (p=0.008 n=5+5)
DeleteAndRestore/Copy-8                 1.19k ± 0%     1.19k ± 0%      ~     (all equal)
DeleteAndRestore/ClearBigFreelist-8      5.00 ± 0%      2.00 ± 0%   -60.00%  (p=0.008 n=5+5)
DeleteAndRestore/Clear-8                1.08k ± 0%     1.08k ± 0%    -0.09%  (p=0.008 n=5+5)
```

```commit-message
```

### Check List

Tests

- Unit test

### Release note

```release-note
None
```

### CODE REVIEW

File: pkg/btree/LICENSE [hunk 1]
```diff
@@ -1,202 +0,0 @@
-
```
Comment (nolouch) [inline pkg/btree/LICENSE:1]: We need keep LICENSE.

File: server/core/rangetree/range_tree.go [hunk 2]
```diff
@@ -110,8 +108,8 @@ func contains(item RangeItem, key []byte) bool {
 
 // Remove removes the given item and return the deleted item.
 func (r *RangeTree) Remove(item RangeItem) RangeItem {
-	if r := r.tree.Delete(item); r != nil {
-		return r.(RangeItem)
+	if r, _ := r.tree.Delete(item); r != nil {
```
Comment (bufferflies) [inline server/core/rangetree/range_tree.go:111]:
```suggestion
	if r, _ := r.tree.Delete(item); r != nil {
	r,_:= r.tree.Delete(item); 
	return r;
```

File: server/core/rangetree/range_tree.go [hunk 3]
```diff
@@ -163,5 +161,5 @@ func (r *RangeTree) GetWithIndex(item RangeItem) (RangeItem, int) {
 	if rst == nil {
```
Comment (bufferflies) [inline server/core/rangetree/range_tree.go:161]: maybe return rst,index rediretly?

File: server/core/region_tree.go [hunk 4]
```diff
@@ -224,17 +255,36 @@ func (t *regionTree) scanRanges() []*RegionInfo {
 
 func (t *regionTree) getAdjacentRegions(region *RegionInfo) (*regionItem, *regionItem) {
 	item := &regionItem{RegionInfo: &RegionInfo{meta: &metapb.Region{StartKey: region.GetStartKey()}}}
-	prevItem, nextItem := t.tree.GetAdjacentItem(item)
+	prevItem, nextItem := t.getAdjacentItem(item)
```
Comment (bufferflies) [inline server/core/region_tree.go:258]: maybe return the prevItem, nextItem directly?

File: pkg/btree/LICENSE (hunk 1 above)
Comment (rleungx) [reply pkg/btree/LICENSE:1]: The LICENSE is contained in the go file.

**Review by bufferflies — APPROVED** [review]

**Review by HunDunDM — APPROVED** [review]
need to delete `pkg/rangetree`?

**Comment by rleungx** [comment]
> need to delete `pkg/rangetree`?

Yes, it has already been deleted.

**Pull request merged**